File: apps/api/app/schemas/meetings.py
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field, field_validator


class UserMinimal(BaseModel):
    id: UUID
    full_name: str
    email: str
    role: str

    model_config = ConfigDict(from_attributes=True)


class MeetingParticipantRead(BaseModel):
    id: UUID
    meeting_id: UUID
    user_id: UUID
    response_status: str
    user: UserMinimal

    model_config = ConfigDict(from_attributes=True)


class MeetingCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    start_at: datetime
    end_at: datetime
    meeting_link: str | None = None
    location: str | None = None
    participants: list[UUID] = Field(..., min_items=1)

    @field_validator("end_at")
    @classmethod
    def validate_end_time(cls, v: datetime, info) -> datetime:
        start = info.data.get("start_at")
        if start and v <= start:
            raise ValueError("End time must be after the start time.")
        return v

    @field_validator("meeting_link")
    @classmethod
    def validate_link(cls, v: str | None) -> str | None:
        if v and not (v.startswith("http://") or v.startswith("https://")):
            raise ValueError("Meeting link must be a valid HTTP or HTTPS URL.")
        return v


class MeetingUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    meeting_link: str | None = None
    location: str | None = None
    participants: list[UUID] | None = None

    @field_validator("meeting_link")
    @classmethod
    def validate_link(cls, v: str | None) -> str | None:
        if v and not (v.startswith("http://") or v.startswith("https://")):
            raise ValueError("Meeting link must be a valid HTTP or HTTPS URL.")
        return v


class MeetingRead(BaseModel):
    id: UUID
    title: str
    description: str | None
    start_at: datetime
    end_at: datetime
    meeting_link: str | None
    location: str | None
    status: str
    organizer_id: UUID
    organizer: UserMinimal
    participants: list[MeetingParticipantRead]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MeetingRespond(BaseModel):
    response_status: str

    @field_validator("response_status")
    @classmethod
    def validate_response_status(cls, v: str) -> str:
        v = v.lower().strip()
        if v not in ["accepted", "declined", "pending"]:
            raise ValueError("Response status must be accepted, declined, or pending.")
        return v

File: apps/web/src/lib/api/meetings.ts
import apiClient from './client';

export interface UserMinimal {
  id: string;
  full_name: string;
  email: string;
  role: string;
  avatar_url?: string | null;
}

export interface MeetingParticipant {
  id: string;
  meeting_id: string;
  user_id: string;
  response_status: 'accepted' | 'declined' | 'pending';
  user: UserMinimal;
}

export interface Meeting {
  id: string;
  title: string;
  description: string | null;
  start_at: string;
  end_at: string;
  meeting_link: string | null;
  location: string | null;
  status: 'scheduled' | 'cancelled';
  organizer_id: string;
  organizer: UserMinimal;
  participants: MeetingParticipant[];
  created_at: string;
  updated_at: string;
}

export interface MeetingCreateInput {
  title: string;
  description?: string;
  start_at: string;
  end_at: string;
  meeting_link?: string;
  location?: string;
  participants: string[]; // UUID strings
}

export interface MeetingUpdateInput {
  title?: string;
  description?: string;
  start_at?: string;
  end_at?: string;
  meeting_link?: string;
  location?: string;
  participants?: string[];
}

export const meetingsApi = {
  getMeetings: async (scope?: 'all'): Promise<Meeting[]> => {
    const response = await apiClient.get<Meeting[]>('/meetings', {
      params: scope ? { scope } : {},
    });
    return response.data;
  },

  getUpcomingMeetings: async (): Promise<Meeting[]> => {
    const response = await apiClient.get<Meeting[]>('/meetings/upcoming');
    return response.data;
  },

  getTodayMeetings: async (scope?: 'all'): Promise<Meeting[]> => {
    const response = await apiClient.get<Meeting[]>('/meetings/today', {
      params: scope ? { scope } : {},
    });
    return response.data;
  },

  getMeeting: async (meetingId: string): Promise<Meeting> => {
    const response = await apiClient.get<Meeting>(`/meetings/${meetingId}`);
    return response.data;
  },

  createMeeting: async (payload: MeetingCreateInput): Promise<Meeting> => {
    const response = await apiClient.post<Meeting>('/meetings', payload);
    return response.data;
  },

  updateMeeting: async (meetingId: string, payload: MeetingUpdateInput): Promise<Meeting> => {
    const response = await apiClient.patch<Meeting>(`/meetings/${meetingId}`, payload);
    return response.data;
  },

  cancelMeeting: async (meetingId: string): Promise<Meeting> => {
    const response = await apiClient.post<Meeting>(`/meetings/${meetingId}/cancel`);
    return response.data;
  },

  respondMeeting: async (meetingId: string, responseStatus: 'accepted' | 'declined' | 'pending'): Promise<Meeting> => {
    const response = await apiClient.post<Meeting>(`/meetings/${meetingId}/respond`, {
      response_status: responseStatus,
    });
    return response.data;
  },
};
